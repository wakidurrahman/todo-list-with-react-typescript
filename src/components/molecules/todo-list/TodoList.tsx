/**
 * TodoList Component
 *
 * A React component that displays a list of todo items with loading and empty states.
 * Features include:
 * - Loading spinner while data is being fetched
 * - Empty state message when no todos exist
 * - List of TodoItem components with consistent spacing
 * - Passes through edit, delete and toggle complete handlers
 * - Responsive layout with Bootstrap
 * - Accessibility features
 *
 * @component
 * @example
 * ```tsx
 * <TodoList
 *   loading={false}
 *   todos={todoItems}
 *   onEdit={handleEdit}
 *   onDelete={handleDelete}
 *   onToggleComplete={handleToggleComplete}
 * />
 * ```
 *
 * @param {Object} props - Component props
 * @param {boolean} props.loading - Loading state flag
 * @param {Todo[]} props.todos - Array of todo items to display
 * @param {Function} props.onEdit - Callback when edit is triggered for a todo
 * @param {Function} props.onDelete - Callback when delete is triggered for a todo
 * @param {Function} props.onToggleComplete - Callback when completion is toggled for a todo
 */

import { Todo } from '../../../types/todo.types';
import TodoItem from '../../atoms/todo-item/TodoItem';

type TodoListProps = {
  loading: boolean;
  todos: Todo[];
  onEdit: (todo: Todo) => void;
  onDelete: (id: string) => void;
  onToggleComplete: (id: string) => void;
};

function TodoList({
  loading,
  todos,
  onEdit,
  onDelete,
  onToggleComplete,
}: TodoListProps) {
  if (loading) {
    return (
      <div className="d-flex justify-content-center mb-4">
        <div className="spinner-border" role="status">
          <span className="visually-hidden">Loading...</span>
        </div>
      </div>
    );
  }

  if (!loading && todos.length === 0) {
    return (
      <div className="alert alert-info" role="alert">
        No todos yet. Add your first todo!
      </div>
    );
  }

  return (
    <div className="d-flex flex-column gap-3">
      {todos.map((todo) => (
        <TodoItem
          key={todo.id}
          todo={todo}
          onEdit={onEdit}
          onDelete={onDelete}
          onToggleComplete={onToggleComplete}
        />
      ))}
    </div>
  );
}

export default TodoList;
