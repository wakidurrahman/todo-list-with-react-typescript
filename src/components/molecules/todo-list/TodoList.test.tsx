import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import TodoList from './TodoList';

describe('TodoList Component', () => {
  const mockTodos = [
    {
      id: '1',
      title: 'First Todo',
      description: 'First Description',
      createdAt: new Date('2024-02-16T00:00:00.000Z'),
      updatedAt: new Date('2024-02-16T00:00:00.000Z'),
    },
    {
      id: '2',
      title: 'Second Todo',
      description: 'Second Description',
      createdAt: new Date('2024-02-16T01:00:00.000Z'),
      updatedAt: new Date('2024-02-16T01:00:00.000Z'),
    },
  ];

  // Snapshot Tests
  describe('Snapshots', () => {
    it('renders list with todos correctly', () => {
      const { container } = render(
        <TodoList todos={mockTodos} onEdit={() => {}} onDelete={() => {}} />
      );
      expect(container).toMatchSnapshot();
    });

    it('renders empty list correctly', () => {
      const { container } = render(
        <TodoList todos={[]} onEdit={() => {}} onDelete={() => {}} />
      );
      expect(container).toMatchSnapshot();
    });
  });

  // Unit Tests
  describe('Functionality', () => {
    const user = userEvent.setup();

    it('displays all todos in the list', () => {
      render(
        <TodoList todos={mockTodos} onEdit={() => {}} onDelete={() => {}} />
      );

      mockTodos.forEach((todo) => {
        expect(screen.getByText(todo.title)).toBeInTheDocument();
        expect(screen.getByText(todo.description)).toBeInTheDocument();
      });
    });

    it('calls onEdit with correct todo when edit button is clicked', async () => {
      const handleEdit = jest.fn();
      render(
        <TodoList todos={mockTodos} onEdit={handleEdit} onDelete={() => {}} />
      );

      const editButtons = screen.getAllByRole('button', { name: /edit/i });
      await user.click(editButtons[0]);

      expect(handleEdit).toHaveBeenCalledWith(mockTodos[0]);
    });

    it('calls onDelete with correct id when delete button is clicked', async () => {
      const handleDelete = jest.fn();
      render(
        <TodoList todos={mockTodos} onEdit={() => {}} onDelete={handleDelete} />
      );

      const deleteButtons = screen.getAllByRole('button', { name: /delete/i });
      await user.click(deleteButtons[1]);

      expect(handleDelete).toHaveBeenCalledWith(mockTodos[1].id);
    });

    it('displays "No todos found" message when list is empty', () => {
      render(<TodoList todos={[]} onEdit={() => {}} onDelete={() => {}} />);

      expect(
        screen.getByText(/no todos yet. add your first todo!/i)
      ).toBeInTheDocument();
    });

    // it('renders todos in correct order (newest first)', () => {
    //   render(
    //     <TodoList todos={mockTodos} onEdit={() => {}} onDelete={() => {}} />
    //   );

    //   const todoElements = screen.getAllByTestId('todo-card');
    //   expect(todoElements[0]).toHaveTextContent(mockTodos[0].title);
    //   expect(todoElements[1]).toHaveTextContent(mockTodos[1].title);
    // });
  });
});
