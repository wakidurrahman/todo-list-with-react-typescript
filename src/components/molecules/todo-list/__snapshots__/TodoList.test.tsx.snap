// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Component Snapshots renders empty list correctly 1`] = `
<div>
  <div
    class="alert alert-info"
    role="alert"
  >
    No todos yet. Add your first todo!
  </div>
</div>
`;

exports[`TodoList Component Snapshots renders list with todos correctly 1`] = `
<div>
  <div
    class="d-flex flex-column gap-1"
  >
    <div
      class="card mb-1"
      data-testid="todo-item-1"
    >
      <div
        class="card-body"
      >
        <div
          class="row align-items-start g-2"
        >
          <div
            class="col"
          >
            <div
              class="form-check"
            >
              <input
                class="form-check-input mt-2"
                id="todo-1"
                type="checkbox"
                value="1"
              />
              <label
                class="form-check-label w-100 "
                for="todo-1"
              >
                <div
                  class="ms-2"
                >
                  <div
                    class="d-flex align-items-center gap-2 mb-3"
                  >
                    <h5
                      class="card-title mb-0 text-break "
                    >
                      First Todo
                    </h5>
                    <span
                      class="badge bg-undefined"
                      title="Priority: undefined"
                    />
                  </div>
                  <h6
                    class="card-subtitle text-body-secondary mb-2"
                  >
                    <i
                      aria-hidden="true"
                      class="bi bi-clock me-1"
                    />
                    Updated:
                     
                    2/16/2024, 9:00:00 AM
                  </h6>
                  <p
                    class="card-text text-break mt-2 mb-0 "
                  >
                    First Description
                  </p>
                </div>
              </label>
            </div>
          </div>
          <div
            class="col-auto"
          >
            <div
              class="btn-group"
            >
              <button
                aria-label="Edit todo "First Todo""
                class="btn btn-secondary btn-sm"
                data-testid="edit-todo-1"
                type="button"
              >
                <i
                  aria-hidden="true"
                  class="bi bi-pencil-square me-1"
                />
                Edit
              </button>
              <button
                aria-label="Delete todo "First Todo""
                class="btn btn-danger btn-sm"
                data-testid="delete-todo-1"
                type="button"
              >
                <i
                  aria-hidden="true"
                  class="bi bi-trash me-1"
                />
                Delete
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card mb-1"
      data-testid="todo-item-2"
    >
      <div
        class="card-body"
      >
        <div
          class="row align-items-start g-2"
        >
          <div
            class="col"
          >
            <div
              class="form-check"
            >
              <input
                class="form-check-input mt-2"
                id="todo-2"
                type="checkbox"
                value="2"
              />
              <label
                class="form-check-label w-100 "
                for="todo-2"
              >
                <div
                  class="ms-2"
                >
                  <div
                    class="d-flex align-items-center gap-2 mb-3"
                  >
                    <h5
                      class="card-title mb-0 text-break "
                    >
                      Second Todo
                    </h5>
                    <span
                      class="badge bg-undefined"
                      title="Priority: undefined"
                    />
                  </div>
                  <h6
                    class="card-subtitle text-body-secondary mb-2"
                  >
                    <i
                      aria-hidden="true"
                      class="bi bi-clock me-1"
                    />
                    Updated:
                     
                    2/16/2024, 10:00:00 AM
                  </h6>
                  <p
                    class="card-text text-break mt-2 mb-0 "
                  >
                    Second Description
                  </p>
                </div>
              </label>
            </div>
          </div>
          <div
            class="col-auto"
          >
            <div
              class="btn-group"
            >
              <button
                aria-label="Edit todo "Second Todo""
                class="btn btn-secondary btn-sm"
                data-testid="edit-todo-2"
                type="button"
              >
                <i
                  aria-hidden="true"
                  class="bi bi-pencil-square me-1"
                />
                Edit
              </button>
              <button
                aria-label="Delete todo "Second Todo""
                class="btn btn-danger btn-sm"
                data-testid="delete-todo-2"
                type="button"
              >
                <i
                  aria-hidden="true"
                  class="bi bi-trash me-1"
                />
                Delete
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
